proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,468 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// text memcache protocol parser based on https://github.com/memcached/memcached/blob/master/doc/protocol.txt
+
+package textmemcache
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// TextMemcacheRule matches against memcached requests
+type TextMemcacheRule struct {
+	// opCode group name
+	command string
+
+	keyExact  string
+	keyPrefix string
+	keyRegex  string
+
+	// allowed commands
+	commands memcacheCommandSet
+	// compiled regex
+	keyExactBytes  []byte
+	keyPrefixBytes []byte
+	regex          *regexp.Regexp
+}
+
+type textMemcacheMeta struct {
+	command []byte
+	keys    [][]byte
+}
+
+// Matches returns true if the TextMemcacheRule matches
+func (rule *TextMemcacheRule) Matches(data interface{}) bool {
+	log.Infof("textmemcache checking rule %v", *rule)
+
+	packetMeta := data.(textMemcacheMeta)
+
+	if !rule.matchCommand(string(packetMeta.command)) {
+		return false
+	}
+
+	if rule.keyExact != "" {
+		for _, key := range packetMeta.keys {
+			if !bytes.Equal(rule.keyExactBytes, key) {
+				return false
+			}
+		}
+		return true
+	}
+
+	if rule.keyPrefix != "" {
+		for _, key := range packetMeta.keys {
+			if !bytes.HasPrefix(key, rule.keyPrefixBytes) {
+				return false
+			}
+		}
+		return true
+	}
+
+	if rule.keyRegex != "" {
+		for _, key := range packetMeta.keys {
+			if !rule.regex.Match(key) {
+				return false
+			}
+		}
+		return true
+	}
+
+	log.Infof("No key rule specified, matching by opcode")
+	return true
+}
+
+func (rule *TextMemcacheRule) matchCommand(cmd string) bool {
+	_, ok := rule.commands[cmd]
+	return ok
+}
+
+// L7TextMemcacheRuleParser parses protobuf L7 rules to and array of TextMemcacheRule
+// May panic
+func L7TextMemcacheRuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		proxylib.ParseError("Can't get L7 rules", rule)
```

Comment:
Shouldn't we `return nil` here?

Reply:
ParseError wraps a panic() that takes care of NACKing the policy updata. Besides, nil would be undestrood as an empty slice, which means wildcard, i.e. allow all L7.